src/Accelerators/NNPA/Conversion/ONNXToZHigh/RewriteONNXForZHigh.cpp

```diff
@@ -170,6 +307,48 @@ void RewriteONNXForZHighPass::runOnOperation() {
                  isUniBroadcatableFirstToSecond(op.B(), op.A())));
   });
 
+  // Legalize MatMulOp if
+  // - both inputs are *the same* N-D, N > 3, or
+  // - one input is N-D, N > 3 and the other is 2-D.
+  // Rewrite patterns will be added to turn this MatMulOp into the one where N-D
+  // will become 3-D.
+  target.addDynamicallyLegalOp<ONNXMatMulOp>([](ONNXMatMulOp op) {
+    Type aType = op.A().getType();
+    Type bType = op.B().getType();
+    if (!isRankedShapedType(aType) || !isRankedShapedType(bType))
+      return true;
+
+    // Only support static shape at this moment though the code supports dynamic
+    // shape as well.
+    //
+    // The reason is that lowering (3Dx3D) ONNXMatMul of dynamic shape to NNPA
```

Comment:
Is the error related to reshaping of 4D to 3D?

Reply:
It seems reshaping of 4D to 3D is correct. For example, keeping the code that reshapes 4D to 3D and using 3D ONNXMatMul leads to the correct result. But if we allow lowering 3D ONNXMatMul to NNPA instead of CPU, we got a wrong result. Looks like something is wrong when lowering a 3D ONNXMatMul of dynamic shape to NNPA or when the lowering is combined with other lowerings.